File: src/entities/Player.ts
import Phaser from 'phaser'
import { PLAYER_CONFIG, COLORS } from '../config'
import type { Vector2 } from '../types'

export class Player extends Phaser.Physics.Arcade.Sprite {
  private health: number
  private maxHealth: number
  private speed: number
  private attackDamage: number
  private attackCooldown: number
  private lastAttackTime: number = 0
  private direction: Vector2 = { x: 0, y: 0 }
  private weaponGraphics!: Phaser.GameObjects.Graphics

  constructor(
    scene: Phaser.Scene,
    x: number,
    y: number,
  ) {
    super(scene, x, y)
    
    this.health = PLAYER_CONFIG.health
    this.maxHealth = PLAYER_CONFIG.maxHealth
    this.speed = PLAYER_CONFIG.speed
    this.attackDamage = PLAYER_CONFIG.attackDamage
    this.attackCooldown = PLAYER_CONFIG.attackCooldown

    // Crear sprite del jugador (círculo relleno)
    const graphics = scene.make.graphics({ x: 0, y: 0, add: false })
    graphics.fillStyle(COLORS.player, 1)
    graphics.fillCircle(PLAYER_CONFIG.size / 2, PLAYER_CONFIG.size / 2, PLAYER_CONFIG.size / 2)
    graphics.generateTexture('player_texture', PLAYER_CONFIG.size, PLAYER_CONFIG.size)
    graphics.destroy()

    this.setTexture('player_texture')
    this.setBounce(0.2)
    this.setDrag(0.99)
    this.setMaxVelocity(this.speed, this.speed)

    scene.physics.world.enable(this)
    scene.add.existing(this)
  }

  update(cursors: Phaser.Input.Keyboard.CursorKeys | null, pointerX: number, pointerY: number) {
    // Reset velocidad
    this.setVelocity(0, 0)

    // Movimiento con WASD
    let moving = false
    const moveSpeed = this.speed

    if (cursors?.up.isDown || (this.scene.input.keyboard?.isDown('W') ?? false)) {
      this.setVelocityY(-moveSpeed)
      moving = true
    }
    if (cursors?.down.isDown || (this.scene.input.keyboard?.isDown('S') ?? false)) {
      this.setVelocityY(moveSpeed)
      moving = true
    }
    if (cursors?.left.isDown || (this.scene.input.keyboard?.isDown('A') ?? false)) {
      this.setVelocityX(-moveSpeed)
      moving = true
    }
    if (cursors?.right.isDown || (this.scene.input.keyboard?.isDown('D') ?? false)) {
      this.setVelocityX(moveSpeed)
      moving = true
    }

    // Normalizar diagonal movement
    if (this.body?.velocity) {
      const velocity = this.body.velocity as Phaser.Math.Vector2
      if (Math.abs(velocity.x) > 0 && Math.abs(velocity.y) > 0) {
        velocity.normalize().scale(moveSpeed)
      }
    }

    // Mirar hacia el mouse
    const dx = pointerX - this.x
    const dy = pointerY - this.y
    const angle = Math.atan2(dy, dx)
    this.setRotation(angle)
  }

  damage(amount: number) {
    this.health -= amount
    if (this.health < 0) this.health = 0
    return this.health <= 0
  }

  getHealth(): number {
    return this.health
  }

  getMaxHealth(): number {
    return this.maxHealth
  }

  canAttack(): boolean {
    return this.scene.time.now - this.lastAttackTime > this.attackCooldown * 1000
  }

  performAttack(): void {
    this.lastAttackTime = this.scene.time.now
  }

  getAttackDamage(): number {
    return this.attackDamage
  }

  getAttackCooldown(): number {
    return this.attackCooldown
  }

  getLastAttackTime(): number {
    return this.lastAttackTime
  }
}

File: src/player.rs
use macroquad::prelude::*;

const PLAYER_COLOR: Color = Color::new(1.0, 0.42, 0.61, 1.0);
const GAME_WIDTH: f32 = 800.0;
const GAME_HEIGHT: f32 = 600.0;

pub struct Player {
    pub x: f32,
    pub y: f32,
    pub health: f32,
    pub max_health: f32,
    pub speed: f32,
    pub size: f32,
    pub attack_damage: f32,
    pub attack_cooldown: f32,
    last_attack_time: f32,
}

impl Player {
    pub fn new(x: f32, y: f32) -> Self {
        Self {
            x,
            y,
            health: 100.0,
            max_health: 100.0,
            speed: 200.0,
            size: 20.0,
            attack_damage: 10.0,
            attack_cooldown: 0.3,
            last_attack_time: -1.0,
        }
    }

    pub fn update(&mut self) {
        let mut dx = 0.0;
        let mut dy = 0.0;

        // Movement
        if is_key_down(KeyCode::W) || is_key_down(KeyCode::Up) {
            dy -= self.speed;
        }
        if is_key_down(KeyCode::S) || is_key_down(KeyCode::Down) {
            dy += self.speed;
        }
        if is_key_down(KeyCode::A) || is_key_down(KeyCode::Left) {
            dx -= self.speed;
        }
        if is_key_down(KeyCode::D) || is_key_down(KeyCode::Right) {
            dx += self.speed;
        }

        // Normalize diagonal movement
        let magnitude = (dx * dx + dy * dy).sqrt();
        if magnitude > 0.0 {
            dx = (dx / magnitude) * self.speed;
            dy = (dy / magnitude) * self.speed;
        }

        let dt = get_frame_time();
        self.x = (self.x + dx * dt).max(self.size).min(GAME_WIDTH - self.size);
        self.y = (self.y + dy * dt).max(self.size).min(GAME_HEIGHT - self.size);
    }

    pub fn draw(&self) {
        let mouse_x = mouse_position().0;
        let mouse_y = mouse_position().1;
        let angle = ((mouse_y - self.y).atan2(mouse_x - self.x)) as f32;

        draw_circle(self.x, self.y, self.size, PLAYER_COLOR);

        // Draw rotation indicator
        let indicator_x = self.x + angle.cos() * (self.size + 5.0);
        let indicator_y = self.y + angle.sin() * (self.size + 5.0);
        draw_circle(indicator_x, indicator_y, 3.0, WHITE);
    }

    pub fn damage(&mut self, amount: f32) {
        self.health = (self.health - amount).max(0.0);
    }

    pub fn can_attack(&self) -> bool {
        get_time() - self.last_attack_time as f64 > self.attack_cooldown as f64
    }

    pub fn perform_attack(&mut self) {
        self.last_attack_time = get_time() as f32;
    }
}
